tap_aggregator/src/aggregator/v1.rs

```diff
@@ -0,0 +1,350 @@
+
+pub fn check_and_aggregate_receipts(
+    domain_separator: &Eip712Domain,
+    receipts: &[Eip712SignedMessage<Receipt>],
+    previous_rav: Option<Eip712SignedMessage<ReceiptAggregateVoucher>>,
+    wallet: &PrivateKeySigner,
+    accepted_addresses: &HashSet<Address>,
+) -> Result<Eip712SignedMessage<ReceiptAggregateVoucher>> {
+    check_signatures_unique(receipts)?;
+
+    // Check that the receipts are signed by an accepted signer address
+    receipts.par_iter().try_for_each(|receipt| {
+        check_signature_is_from_one_of_addresses(receipt, domain_separator, accepted_addresses)
+    })?;
+
+    // Check that the previous rav is signed by an accepted signer address
+    if let Some(previous_rav) = &previous_rav {
+        check_signature_is_from_one_of_addresses(
+            previous_rav,
+            domain_separator,
+            accepted_addresses,
+        )?;
+    }
+
+    // Check that the receipts timestamp is greater than the previous rav
+    check_receipt_timestamps(receipts, previous_rav.as_ref())?;
+
+    // Get the allocation id from the first receipt, return error if there are no receipts
+    let allocation_id = match receipts.first() {
+        Some(receipt) => receipt.message.allocation_id,
+        None => return Err(tap_core::Error::NoValidReceiptsForRavRequest.into()),
+    };
+
+    // Check that the receipts all have the same allocation id
+    check_allocation_id(receipts, allocation_id)?;
+
+    // Check that the rav has the correct allocation id
+    if let Some(previous_rav) = &previous_rav {
+        let prev_id = previous_rav.message.allocationId;
+        if prev_id != allocation_id {
+            return Err(tap_core::Error::RavAllocationIdMismatch {
+                prev_id: format!("{prev_id:#X}"),
+                new_id: format!("{allocation_id:#X}"),
+            }
+            .into());
+        }
+    }
+
+    // Aggregate the receipts
+    let rav = ReceiptAggregateVoucher::aggregate_receipts(allocation_id, receipts, previous_rav)?;
+
+    // Sign the rav and return
+    Ok(Eip712SignedMessage::new(domain_separator, rav, wallet)?)
+}
+
+fn check_signature_is_from_one_of_addresses<M: SolStruct>(
+    message: &Eip712SignedMessage<M>,
+    domain_separator: &Eip712Domain,
+    accepted_addresses: &HashSet<Address>,
+) -> Result<()> {
+    let recovered_address = message.recover_signer(domain_separator)?;
+    if !accepted_addresses.contains(&recovered_address) {
+        bail!(tap_core::Error::InvalidRecoveredSigner {
+            address: recovered_address,
+        });
+    }
+    Ok(())
+}
+
+fn check_allocation_id(
+    receipts: &[Eip712SignedMessage<Receipt>],
+    allocation_id: Address,
+) -> Result<()> {
+    for receipt in receipts.iter() {
+        let receipt = &receipt.message;
+        if receipt.allocation_id != allocation_id {
+            return Err(tap_core::Error::RavAllocationIdNotUniform.into());
+        }
+    }
+    Ok(())
+}
```

Comment:
It would be nice to fix returning errors where we mean a `bool` but I see you're just moving code around.